File: cpp_generator/Parser.cpp
#include <iostream>
#include <string>
#include "reflect/Parser.hpp"
#include "reflect/Utils.hpp"

namespace reflect::parser
{
    void ParsedField::Debug() const
    {
    }

    void ParsedFunction::Debug() const
    {
        std::cout << "Function: " << result << " " << name << "(";
        for (auto& arg : arguments)
        {
            std::cout << arg->type << " " << arg->name;
            if (arg != arguments.at(arguments.size() - 1))
            {
                std::cout << ",";
            }
        }

        std::cout << ")" << std::endl;
    }

    void ParsedProperty::Debug() const
    {
        ParsedField::Debug();
    }

    void ParsedType::Debug() const
    {
        switch (type)
        {
        case Class:
            std::cout << "Class : " << name << std::endl;
            break;
        case Enum:
            std::cout << "Enum : " << name << std::endl;
            break;
        case Struct:
            std::cout << "Struct : " << name << std::endl;
            break;
        default:
            std::cout << "Unknown : " << name << std::endl;
            break;
        }
    }

    void ParsedStruct::Debug() const
    {
        ParsedType::Debug();
        for (auto& f : fields)
        {
            f->Debug();
        }
    }

    void ParsedClass::Debug() const
    {
        ParsedStruct::Debug();
    }

    void ParsedFile::Debug() const
    {
        std::cout << "File: " << filePath << "\n";
        for (auto& t : types)
        {
            t->Debug();
        }
    }

    FileParser::FileParser(const std::filesystem::path& path)
    {
        filePath = path;
        stream.open(path);
    }

    bool FileParser::InternalReadLine(std::string& line)
    {
        const auto result = static_cast<bool>(std::getline(stream, line));
        if (result)
        {
            ++lineNumber;
        }
        return result;
    }

    bool FileParser::ReadLine(std::string& line)
    {
        std::string lineTemp;
        auto result = false;
        do
        {
            if (utils::trim(lineTemp).empty())
            {
                result = InternalReadLine(lineTemp);
            }

            if (_isInComment)
            {
                if (const auto pos = lineTemp.find("*/"); pos != std::string::npos)
                {
                    lineTemp = lineTemp.substr(pos + 2);
                    _isInComment = false;
                }
                else
                {
                    result = InternalReadLine(lineTemp);
                }
            }
            else if (const auto pos = lineTemp.find("/*"); pos != std::string::npos)
            {
                lineTemp = lineTemp.substr(pos + 2);
                _isInComment = true;
            }
            else if (const auto pos = lineTemp.find("//"); pos != std::string::npos)
            {
                lineTemp = lineTemp.substr(pos);
            }
        }
        while (result && (utils::trim(lineTemp).empty() || _isInComment));

        line = lineTemp;

        return result;
    }

    std::shared_ptr<ParsedProperty> FileParser::ParseProperty()
    {
        auto result = std::make_shared<ParsedProperty>();

        std::string line;
        ReadLine(line);
        line = utils::trim(line);

        std::smatch match;
        if (std::regex_match(line, match, REFLECT_PROPERTY_REGEX))
        {
            result->type = utils::trim(std::string(match[1].str()));
            result->name = utils::trim(std::string(match[2].str()));
        }

        return result;
    }

    std::shared_ptr<ParsedFunction> FileParser::ParseFunction()
    {
        auto result = std::make_shared<ParsedFunction>();

        std::string line;
        ReadLine(line);
        line = utils::trim(line);
        const auto staticStr = std::string("static ");
        
        result->bIsStatic = line.find(staticStr) != std::string::npos;
        if(result->bIsStatic)
        {
            auto strSize = staticStr.size();
            line = line.substr(line.find(staticStr) + strSize);
        }
        const auto splitLine = utils::split(utils::trim(line), " ");
        result->name = utils::split(splitLine.at(1), "(").at(0);
        result->result = splitLine.at(0);
        auto argsLine = utils::trim(line);
        argsLine = utils::split(argsLine, "(").at(1);
        size_t pos = 0;
        pos = argsLine.find(")");
        while (pos == std::string::npos)
        {
            ReadLine(line);
            argsLine += utils::trim(line);
            pos = argsLine.find(")");
        }
        pos = argsLine.find(")");
        argsLine = argsLine.substr(0, pos);
        const auto argsArr = utils::split(utils::trim(argsLine), ",");

        for (auto& arg : argsArr)
        {
            std::smatch match;
            if (std::regex_match(arg, match, REFLECT_ARGUMENT_REGEX))
            {
                auto type = match[1];
                auto name = match[2];
                result->arguments.push_back(
                    std::make_shared<ParsedFunctionArgument>(utils::trim(type.str()), utils::trim(name.str())));
            }
        }
        return result;
    }

    std::shared_ptr<ParsedClass> FileParser::ParseClass()
    {
        auto result = std::make_shared<ParsedClass>();
        std::string line;
        const std::string searchStr = "class ";
        while (line.find(searchStr) == std::string::npos)
        {
            ReadLine(line);
            line = utils::trim(line);
        }
        // Get Class Name
        const auto classSpecPos = line.find(searchStr);

        if (classSpecPos == std::string::npos) return {};

        const auto splitResult = utils::split(utils::trim(line.substr(classSpecPos + searchStr.size())), " ");

        if (splitResult.empty()) return {};

        result->name = splitResult.at(0);

        // Parse Fields
        while (ReadLine(line))
        {
            if (line.find("};") != std::string::npos) return result;

            if (line.find(REFLECT_PROPERTY_MACRO) != std::string::npos)
            {
                if (auto r = ParseProperty())
                {
                    result->fields.push_back(r);
                }
            }
            else if (line.find(REFLECT_FUNCTION_MACRO) != std::string::npos)
            {
                if (auto r = ParseFunction())
                {
                    result->fields.push_back(r);
                }
            }
        }

        return result;
    }

    std::shared_ptr<ParsedStruct> FileParser::ParseStruct()
    {
        auto result = std::make_shared<ParsedStruct>();
        std::string line;
        const std::string searchStr = "struct ";
        while (line.find(searchStr) == std::string::npos)
        {
            ReadLine(line);
            line = utils::trim(line);
        }
        // Get Class Name
        const auto classSpecPos = line.find(searchStr);

        if (classSpecPos == std::string::npos) return {};

        const auto splitResult = utils::split(utils::trim(line.substr(classSpecPos + searchStr.size())), " ");

        if (splitResult.empty()) return {};

        result->name = splitResult.at(0);

        // Parse Fields
        while (ReadLine(line))
        {
            if (line.find("};") != std::string::npos) return result;

            if (line.find(REFLECT_PROPERTY_MACRO) != std::string::npos)
            {
                if (auto r = ParseProperty())
                {
                    result->fields.push_back(r);
                }
            }
            else if (line.find(REFLECT_FUNCTION_MACRO) != std::string::npos)
            {
                if (auto r = ParseFunction())
                {
                    result->fields.push_back(r);
                }
            }
        }

        return result;
    }

    std::shared_ptr<ParsedFile> FileParser::Parse()
    {
        if (!stream.is_open()) return {};

        auto result = std::make_shared<ParsedFile>();

        result->filePath = filePath;
        std::string line;
        std::string content;
        while (ReadLine(line))
        {
            if (line.find(REFLECT_CLASS_MACRO) != std::string::npos)
            {
                if (auto cls = ParseClass())
                {
                    result->types.push_back(cls);
                }
            }
            else if (line.find(REFLECT_STRUCT_MACRO) != std::string::npos)
            {
                if (auto rstruct = ParseStruct())
                {
                    result->types.push_back(rstruct);
                }
            }
        }

        lineNumber = 0;
        stream.close();

        return result;
    }


    std::shared_ptr<ParsedFile> Parser::ParseFile(const std::filesystem::path& path) const
    {
        if (!std::filesystem::exists(path)) return {};

        FileParser parser(path);

        return parser.Parse();
    }

    std::vector<std::shared_ptr<ParsedFile>> Parser::Parse() const
    {
        std::vector<std::shared_ptr<ParsedFile>> result;

        for (auto& file : files)
        {
            std::cout << "Parsing " << file << std::endl;
            if (auto r = ParseFile(file); !r->types.empty())
            {
                result.push_back(r);
            }
            std::cout << "Parsed " << file << std::endl;
        }
        return result;
    }
}

File: python/reflect/parser.py
import re
from io import TextIOWrapper
from typing import Union
from reflect.constants import REFLECT_CLASS_MACRO, REFLECT_STRUCT_MACRO, REFLECT_PROPERTY_MACRO,REFLECT_FUNCTION_MACRO, REFLECT_PROPERTY_REGEX, REFLECT_ARGUMENT_REGEX, REFLECT_FUNCTION_REGEX
from reflect.types import ParsedFile,ParsedClass, ParsedStruct, ParsedFunction, ParsedProperty, ParsedFunctionArgument
class FileParser:
    def __init__(self,file_path: str) -> None:
        self.is_in_comment = False
        self.line_number = 0
        self.file_path = file_path
        self.file: Union[None,TextIOWrapper] = None

    def _read(self) -> Union[None,str]:
        if self.file is None:
            return None
        
        line = self.file.readline()

        if line == "":
            return None

        self.line_number += 1

        return line

    def read(self) -> Union[None,str]:
        
        line = ""

        while True:
            line_temp = self._read()

            if line_temp is None:
                return None
            
            line += line_temp.strip()

            if self.is_in_comment:
                if "*/" in line:
                    line = line[line.index("*/") + 1:]
                    self.is_in_comment = False
                else:
                    line = self._read()

            elif "/*" in line:
                line = line[line.index("/*") + 1:]
                self.is_in_comment = True
            elif "//" in line:
                line = line[:line.index("//")]

            if len(line) == 0 or self.is_in_comment:
                continue

            return line
            


    def parse_tags(self,macro_line:str) -> list[str]:
        line =  macro_line[macro_line.index('(') + 1:len(macro_line[1:]) - macro_line[::-1].index(')')]
        return  list(filter(lambda a : len(a) > 0,map(lambda a : a.strip(),line.strip().split(','))))

    def parse_property(self,macro_iine:str) -> Union[None,ParsedProperty]:
        result  = ParsedProperty(self.parse_tags(macro_iine))

        line = self.read()

        match_result = re.findall(REFLECT_PROPERTY_REGEX,line)

        if len(match_result) < 1 or len(match_result[0]) < 2:
            return None
        
        match_result = match_result[0]
        
        result.type = match_result[0]
        result.name = match_result[1]

        return result



    def parse_function(self,macro_iine:str) -> Union[None,ParsedFunction]:
        line = self.read()
        static_str = "static "

        result = ParsedFunction(self.parse_tags(macro_iine))

        match_result = re.findall(REFLECT_FUNCTION_REGEX,line)

        if len(match_result) < 1 or len(match_result[0]) < 3:
            return None
        
        match_result = match_result[0]
        
        result.is_static = not match_result[0].strip() == ""
        result.return_type = match_result[1]
        result.name = match_result[2]

        args_line = line.split("(")[1]

        while ")" not in args_line:
            args_line += self.read()

        args_line = args_line[:args_line.index(")")]

        args_arr = args_line.split(",")

        for arg in args_arr:
            match_result = re.findall(REFLECT_ARGUMENT_REGEX,arg)

            if len(match_result) < 1 or len(match_result[0]) < 2:
                continue

            match_result = match_result[0]

            p_arg = ParsedFunctionArgument()
            p_arg.name = match_result[1]
            p_arg.type = match_result[0]
            result.arguments.append(p_arg)

        return result

    def parse_class(self,macro_iine:str) -> Union[None,ParsedClass]:
        search_str = "class "
        line = self.read()

        while line is not None and search_str not in line:
            line = self.read()

        if line is None:
            return None
        
        split_result = line[line.index(search_str) + len(search_str):].split(" ")

        if len(split_result) == 0:
            return None
        
        result = ParsedClass(self.parse_tags(macro_iine))
        result.name = split_result[0]

        line = self.read()
        
        while line is not None:
            if "};" in line:
                return result
            
            if REFLECT_PROPERTY_MACRO in line:
                r = self.parse_property(line)

                if r is not None:
                    result.fields.append(r)
            elif REFLECT_FUNCTION_MACRO in line:
                r = self.parse_function(line)

                if r is not None:
                    result.fields.append(r)

            line = self.read()





    def parse_struct(self,macro_iine:str) -> Union[None,ParsedStruct]:
        search_str = "struct "
        line = self.read()

        while line is not None and search_str not in line:
            line = self.read()

        if line is None:
            return None
        
        split_result = line[line.index(search_str) + len(search_str):].split(" ")

        if len(split_result) == 0:
            return None
        
        result = ParsedStruct(self.parse_tags(macro_iine))
        result.name = split_result[0]

        line = self.read()
        
        while line is not None:
            if "};" in line:
                return result
            
            if REFLECT_PROPERTY_MACRO in line:
                r = self.parse_property(line)

                if r is not None:
                    result.fields.append(r)
            elif REFLECT_FUNCTION_MACRO in line:
                r = self.parse_function(line)

                if r is not None:
                    result.fields.append(r)

            line = self.read()

    def parse(self) -> Union[None,ParsedFile]:
        with open(self.file_path,'r') as f:
            self.file = f

            result = ParsedFile()

            result.file_path = self.file_path

            line = self.read()

            while line is not None:
                if REFLECT_CLASS_MACRO in line:
                    r = self.parse_class(line)
                    if r is not None:
                        result.types.append(r)
                elif REFLECT_STRUCT_MACRO in line:
                    r = self.parse_struct(line)
                    if r is not None:
                        result.types.append(r)
                
                line = self.read()

            self.line_number = 0

            return result
        

class Parser:
    def __init__(self,files: list[str]) -> None:
        self.files = files

    def parse_file(self,file_path: str) -> Union[None,ParsedFile]:
        parser = FileParser(file_path=file_path)

        return parser.parse()

    def parse(self) -> list[ParsedFile]:
        results: list[ParsedFile] = []

        for file in self.files:
            print(f"Parsing {file}")

            result = self.parse_file(file)

            if result is not None and len(result.types) > 0:
                results.append(result)

            print(f"Parsed {file}")

        return results